flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/clustering/KMeans.scala

```diff
@@ -0,0 +1,614 @@
+          override def open(parameter: Configuration): Unit ={
+            currentSet = getRuntimeContext.getBroadcastVariable(PARINIT_SET).asScala
+            cost = getRuntimeContext.getBroadcastVariable(PARINIT_COST).get(0)
+            oversamplingFactor = getRuntimeContext.getBroadcastVariable(PARINIT_SAMPLE).get(0)
+            rng = new Random()
+          }
+          override def filter(value: Vector): Boolean = {
+            rng.nextDouble() <
+              oversamplingFactor * Math.pow(findNearestCentroid(value, currentSet)._2, 2) / cost
+          }
+        }).withBroadcastSet(currentCost, PARINIT_COST)
+          .withBroadcastSet(currentSet, PARINIT_SET)
+          .withBroadcastSet(oversamplingFactor, PARINIT_SAMPLE)
+
+        // keep taking unions of independent samples at each step
+        currentSet.union(sampledSet.map(x => LabeledVector(0, x)))
+      }
+    }
+
+    // now assign weights to points in the set
+    val weightedSample = data.mapWithBcSet(unionOfSamples){
+      (vector, sampledSet) => {
+        val samples = sampledSet.toList
+        var minDistance: Double = Double.MaxValue
+        var closestCentroidIndex: Int = -1
+        for (i <- 0 to samples.size - 1) {
+          val distance = EuclideanDistanceMetric().distance(vector, samples(i).vector)
+          if (distance < minDistance) {
+            minDistance = distance
+            closestCentroidIndex = i
+          }
+        }
+        // just assign a label of 1. We'll figure this out later.
+        (closestCentroidIndex, samples(closestCentroidIndex).vector, 1)
+      }
+    }.groupBy(0)
+      .reduce((a, b) => (a._1, a._2, a._3 + b._3))
+      .map(x => (x._2,x._3))
+
+    // finally, do a kmeans++ on this weighted set
+    kmeans(weightedSample, k, blocks)
+  }
+
+  /**
+   * Randomly initializes centroids from the data.
+   * Data is considered to be weighted.
+   *
+   * @param data Training data set
+   * @param k Number of centroids to be picked
+   * @return Initial random centroids
+   */
+  private def random(
+      data: DataSet[(Vector, Int)],
+      k: Int)
+    : DataSet[Seq[LabeledVector]] = {
+    // we'll sample 10 times as many points as we actually need
+    // TODO Modify to use the Random Sample Operator as and when added.
+
+    val fraction = data.map(x => 1).reduce(_ + _).map(x => 10 * (k + 0.0) / x)
+
+    val extraSampledSet = data.filter(new RichFilterFunction[(Vector, Int)] {
+      var rng: Random = _
+      var fraction: Double = _
+
+      override def open(parameters: Configuration): Unit ={
+        rng = new Random()
+        fraction = getRuntimeContext.getBroadcastVariable(RANDOM_FRACTION).get(0)
+      }
+      override def filter(value: (Vector,Int)): Boolean = {
+        rng.nextDouble() < fraction * value._2
+      }}
+    ).withBroadcastSet(fraction, RANDOM_FRACTION).map(x => x._1)
+
+    data.getExecutionEnvironment.fromElements(k).mapWithBcSet(extraSampledSet){
```

Comment:
I don't know whether this is the best way to create a sample of `k` elements. What you basically do is to collect the `extraSampledSet` on one node. I think it would be better to simply use `DataSetUtils.sampleWithSize` here.